jina/importer.py

```diff
@@ -26,62 +92,32 @@ def import_classes(namespace: str, targets=None,
 
     import os, re
 
-    if namespace == 'jina.executors':
-        import_type = 'ExecutorType'
-        if import_once and IMPORTED.executors:
-            return
-    elif namespace == 'jina.drivers':
-        import_type = 'DriverType'
-        if import_once and IMPORTED.drivers:
-            return
-    elif namespace == 'jina.hub':
-        import_type = 'ExecutorType'
-        if import_once and IMPORTED.hub:
-            return
-    else:
+    _namespace2type = {
+        'jina.executors': 'ExecutorType',
+        'jina.drivers': 'DriverType',
+        'jina.hub': 'ExecutorType'
+    }
+    _import_type = _namespace2type.get(namespace)
+    if _import_type is None:
         raise TypeError(f'namespace: {namespace} is unrecognized')
 
-    from setuptools import find_packages
-    import pkgutil
-    from pkgutil import iter_modules
+    _targets = _get_targets(targets)
+    if _targets is None:
+        raise TypeError(f'target must be a set, but received {targets!r}')
 
-    try:
-        path = os.path.dirname(pkgutil.get_loader(namespace).path)
-    except AttributeError:
-        if namespace == 'jina.hub':
-            warnings.warn(f'hub submodule is not initialized. Please try "git submodule update --init"', ImportWarning)
+    _imported_property = namespace.split('.')[-1]
+    _is_imported = getattr(IMPORTED, _imported_property)
+    if import_once and _is_imported:
         return {}
 
-    modules = set()
-
-    for info in iter_modules([path]):
-        if (namespace != 'jina.hub' and not info.ispkg) or (namespace == 'jina.hub' and info.ispkg):
-            modules.add('.'.join([namespace, info.name]))
-
-    for pkg in find_packages(path):
-        modules.add('.'.join([namespace, pkg]))
-        pkgpath = path + '/' + pkg.replace('.', '/')
-        for info in iter_modules([pkgpath]):
-            if (namespace != 'jina.hub' and not info.ispkg) or (namespace == 'jina.hub' and info.ispkg):
-                modules.add('.'.join([namespace, pkg, info.name]))
-
-    # filter
-    ignored_module_pattern = r'\.tests|\.api|\.bump_version'
-    modules = {m for m in modules if not re.findall(ignored_module_pattern, m)}
-
+    modules = _get_modules(namespace)
+    if not modules:
+        return {}
```

Comment:
Shouldn't that deserve a proper logging/exception throwing? Further down the function, we would throw an Exception, if not all `_targets` are handled. But now we don't. Feels strange, even thought I must admit, I did not totally went into the details of this function.

Reply:
A warning message is added.

Reply:
After removing `targets`, the return becomes clearer and there is no longer exception thrown out by the function.